src/metabase/models/card.clj

```diff
@@ -654,12 +654,88 @@
       (pre-update-check-sandbox-constraints changes)
       (assert-valid-type (merge old-card-info changes)))))
 
+(defn- derive-ident [prefix entity_id stage-number tail]
+  (let [entity_id (if (instance? clojure.lang.IDeref entity_id)
+                    @entity_id
+                    entity_id)]
+    (if (= entity_id ::before-insert)
+      ;; If this is a fresh insert, it's safe to generate them at random rather than deriving them!
+      (u/generate-nano-id)
+      ;; If entity_id is provided instead, then derive an opaque string based on that entity_id.
+      (str prefix "_" entity_id "@" stage-number "__" tail))))
+
+(defn- ensure-clause-idents-list [existing xs prefix stage-number {:keys [entity_id] :as _ctx}]
+  (into {} (map (fn [i]
+                  [i (or (get existing i)
+                         (derive-ident prefix entity_id stage-number i))]))
+        (range (count xs))))
+
+(defn- ensure-clause-idents-expressions [existing expressions stage-number {:keys [entity_id] :as _ctx}]
+  (m/map-kv-vals (fn [expr-name _expr-clause]
+                   (or (get existing expr-name)
+                       (derive-ident "expression" entity_id stage-number expr-name)))
+                 expressions))
+
+(defn- ensure-clause-idents-joins [joins stage-number {:keys [entity_id] :as _ctx}]
+  (mapv #(assoc % :ident (derive-ident "join" entity_id stage-number (:alias %)))
+        joins))
+
+(defn- ensure-clause-idents-inner [inner-query ctx]
+  (let [{:keys [query stage-number]} (if-let [source-query (:source-query inner-query)]
+                                       (-> (ensure-clause-idents-inner source-query ctx)
+                                           (update :query #(assoc inner-query :source-query %)))
```

Comment:
I agree it's a bit tricksy, but I also didn't see a cleaner way to do this. A `loop` is hard because it's going the wrong way. It needs to be postorder so that we have an accurate `stage-number` for the body of this `let`.

I've added comments calling out the recursion more explicitly, and I'm open to suggestions that are easier to follow. (The only alternative I can see is a HOF like `map-stages` (cf. `map-indexed`) that abstracts some of the machinery.)

Reply:
I was thinking that using a `let` form instead of threading and having a literal map `{:query ..., :stage-number ...}` as the value would make things clearer. (If we do that, then a pair in the form of a vector) would work too.)

> I've rewritten this to be based on a new util mbql.u/map-stages; WDYT?

I'd have to see the rewritten form for a final call, but I trust your taste.

Reply:
The revised version is posted, PTAL.